In [1]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('FoodBalanceSheets_E_Africa_NOFLAG.csv', encoding = 'latin-1')

In [4]:
df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

In [5]:
df.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

### 1. What is the total Protein supply quantity in Madagascar in 2015?

In [27]:
# What is the total Protein supply quantity in Madagascar in 2015?
df[df.Area == 'Madagascar'].groupby('Element')['Y2015'].sum()

Element
Domestic supply quantity                  31214.98
Export Quantity                             494.75
Fat supply quantity (g/capita/day)           91.85
Feed                                       2070.22
Food                                      21120.65
Food supply (kcal/capita/day)              7685.00
Food supply quantity (kg/capita/yr)         871.59
Import Quantity                            1721.80
Losses                                     2176.00
Other uses (non-food)                      3957.12
Processing                                 1803.00
Production                                29482.89
Protein supply quantity (g/capita/day)      173.05
Residuals                                  -616.00
Seed                                        699.00
Stock Variation                            -505.00
Total Population - Both sexes             24234.00
Name: Y2015, dtype: float64

> The total protein supply quantity in Madagascar 2015 was 173.05

### 2. What is the total sum of Wine produced in 2015 and 2018 respectively?

In [25]:
# What is the total sum of Wine produced in 2015 and 2018 respectively?
# First we find the sum of wine produced in 2015
df.groupby('Item')['Y2015'].sum()

Item
Alcohol, Non-Food        2180.00
Alcoholic Beverages     98783.72
Animal Products         11811.73
Animal fats            200675.72
Apples and products     10559.15
                         ...    
Vegetables, Other      158104.08
Vegetal Products       107064.17
Wheat and products     234710.51
Wine                     4251.81
Yams                   203151.78
Name: Y2015, Length: 119, dtype: float64

> The total amount of wine in 2015 was 4251.81

In [26]:
# Secondly, the sum of wine produced in 2018
df.groupby('Item')['Y2018'].sum()

Item
Alcohol, Non-Food        2293.00
Alcoholic Beverages     97847.27
Animal Products         11578.61
Animal fats            269648.27
Apples and products      9640.51
                         ...    
Vegetables, Other      163987.21
Vegetal Products       107775.39
Wheat and products     242645.19
Wine                     4039.32
Yams                   221272.09
Name: Y2018, Length: 119, dtype: float64

> The total amount of wine in 2018 was 4039.32 

### 3. 
Given the following numpy array 

array  = ([[94, 89, 63],

           [93, 92, 48],

           [92, 94, 56]])

How would you select  the elements in bold and italics from the array?


In [35]:
# Create the array
array = np.array([[94, 89, 63],[93, 92, 48], [92, 94, 56]])
array

array([[94, 89, 63],
       [93, 92, 48],
       [92, 94, 56]])

In [36]:
array[:2, 1:]

array([[89, 63],
       [92, 48]])

### 4. Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?

In [50]:
df.groupby('Area')['Y2017'].sum().sort_values()

Area
Sudan (former)                       0.00
Ethiopia PDR                         0.00
Comoros                             59.84
Seychelles                         442.34
Sao Tome and Principe            12662.63
Cabo Verde                       14650.74
Guinea-Bissau                    19102.77
Lesotho                          21267.96
Botswana                         22101.30
Djibouti                         22729.91
Gambia                           23154.18
Gabon                            27979.64
Liberia                          29342.20
Namibia                          29874.89
Central African Republic         29937.00
Congo                            41181.68
Togo                             49841.88
Mauritius                        51114.83
Eswatini                         54343.33
Sierra Leone                     55311.33
Chad                             71594.68
Rwanda                           73663.69
Zimbabwe                         75919.34
Senegal                      

> The area with the highest sum was Nigeria

### 5. What would be the output for?

S = [['him', 'sell'], [90, 28, 43]]

S[0][1][1]

In [52]:
s = [['him', 'sell'], [90, 28, 43]]
s[0][1][1]

'e'

### 6. Which of these python data structures is unorderly?

In [53]:
print('set')

set


### 7. How would you check for the number of rows and columns in a pandas DataFrame named df?

In [54]:
df.shape

(60943, 12)

### 8. Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?

In [60]:
df.groupby('Area')['Y2017'].sum().sort_values()[:7]

Area
Sudan (former)               0.00
Ethiopia PDR                 0.00
Comoros                     59.84
Seychelles                 442.34
Sao Tome and Principe    12662.63
Cabo Verde               14650.74
Guinea-Bissau            19102.77
Name: Y2017, dtype: float64

> Guinea-Bissau has the the 7th lowest sum in 2017

### 9. Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?

In [73]:
# Check the year 2014
df.groupby('Element')['Y2014'].sum()[-3:]

Element
Stock Variation                    58749.83
Total Population - Both sexes    1031585.00
Tourist consumption                  416.00
Name: Y2014, dtype: float64

In [74]:
# check the year 2015
df.groupby('Element')['Y2015'].sum()[-3:]

Element
Stock Variation                    34910.99
Total Population - Both sexes    1058081.00
Tourist consumption                  349.00
Name: Y2015, dtype: float64

In [76]:
df.groupby('Element')['Y2016'].sum()[-3:]

Element
Stock Variation                    33140.12
Total Population - Both sexes    1085107.00
Tourist consumption                   89.00
Name: Y2016, dtype: float64

In [77]:
df.groupby('Element')['Y2017'].sum()[-3:]

Element
Stock Variation                    54316.91
Total Population - Both sexes    1112641.00
Tourist consumption                   91.00
Name: Y2017, dtype: float64

In [78]:
df.groupby('Element')['Y2018'].sum()[-3:]

Element
Stock Variation                    20577.91
Total Population - Both sexes    1140605.00
Tourist consumption                   90.00
Name: Y2018, dtype: float64

> Year 2014 has the highest sum of stock variation of  58749.83

### 10. Consider the following list of tuples:

y = [(2, 4), (7, 8), (1, 5, 9)]

How would you assign element 8 from the list to a variable x?

In [80]:
# Create the tuple 
y = [(2, 4), (7, 8), (1, 5, 9)]

In [83]:
# assign element 8 to variable x
x = y[1][-1]

In [84]:
x

8

### 11. What is the total number of unique countries in the dataset?

In [88]:
len(df.Area.unique())

49

> There are 49 total unique countries

### 12. What is the total number and percentage of missing data in 2014 to 3 decimal places?

In [103]:
# Check the total number of values in the Y2014 column
len(df['Y2014'])

60943

In [108]:
missing_data_2014 = df['Y2014'].isnull().sum()
percentage = (missing_data_2014/len(df['Y2014'])) * 100
txt = "The total number of missing data is {} and the percentage is {:.4}"
print(txt.format(missing_data_2014, percentage))

The total number of missing data is 1589 and the percentage is 2.607


### 13. Which of the following dataframe methods can be used to access elements across rows and columns?

In [112]:
df.iloc[:]

Area Code      Area  Item Code           Item  Element Code  \
0              4   Algeria       2501     Population           511   
1              4   Algeria       2501     Population          5301   
2              4   Algeria       2901    Grand Total           664   
3              4   Algeria       2901    Grand Total           674   
4              4   Algeria       2901    Grand Total           684   
...          ...       ...        ...            ...           ...   
60938        181  Zimbabwe       2899  Miscellaneous          5142   
60939        181  Zimbabwe       2899  Miscellaneous           645   
60940        181  Zimbabwe       2899  Miscellaneous           664   
60941        181  Zimbabwe       2899  Miscellaneous           674   
60942        181  Zimbabwe       2899  Miscellaneous           684   

                                      Element             Unit     Y2014  \
0               Total Population - Both sexes     1000 persons  38924.00   
1                    Domestic supply quantity      1000 tonnes      0.00   
2               Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3      Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4          Fat supply quantity (g/capita/day)     g/capita/day     80.06   
...                                       ...              ...       ...   
60938                                    Food      1000 tonnes     42.00   
60939     Food supply quantity (kg/capita/yr)               kg      3.06   
60940           Food supply (kcal/capita/day)  kcal/capita/day      3.00   
60941  Protein supply quantity (g/capita/day)     g/capita/day      0.10   
60942      Fat supply quantity (g/capita/day)     g/capita/day      0.04   

          Y2015     Y2016     Y2017     Y2018  
0      39728.00  40551.00  41389.00  42228.00  
1          0.00      0.00      0.00      0.00  
2       3379.00   3372.00   3341.00   3322.00  
3         94.35     94.72     92.82     91.83  
4         79.36     77.40     80.19     77.28  
...         ...       ...       ...       ...  
60938     46.00     33.00     19.00     16.00  
60939      3.33      2.35      1.33      1.08  
60940      4.00      3.00      1.00      1.00  
60941      0.11      0.08      0.04      0.04  
60942      0.05      0.03      0.02      0.01  

[60943 rows x 12 columns]

### 14 . A pandas Dataframe with dimensions (100,3) has how many features and observations?

In [113]:
print('A pandas dataframe has no attribute feature or observation')

A pandas dataframe has no attribute feature or observation


### 15. Which year had the least correlation with ‘Element Code’?

In [117]:
df.corr()['Element Code'].sort_values()

Item Code      -0.024683
Area Code      -0.000209
Y2016           0.023444
Y2015           0.023889
Y2017           0.024254
Y2018           0.024279
Y2014           0.024457
Element Code    1.000000
Name: Element Code, dtype: float64

> Year 2016 had the least correlation with Element Code.

### 16. What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [123]:
mean = df['Y2017'].mean()
std = df['Y2017'].std()
txt = "The mean is {:3f} and the standard deviation is {:3f}"
print(txt.format(mean, std))

The mean is 140.917765 and the standard deviation is 1671.862359


### 17. If you have the following list

lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]

col = [‘Age’,’Nationality’,’Overall’]

How do you create a pandas DataFrame using this list, to look like the table below?

In [126]:
lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]

col = ['Age','Nationality','Overall']

pd.DataFrame(lst, columns = col, index = [1, 2, 3])

Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

In [127]:

pd.DataFrame(lst, columns = col, index = [i for i in range(1,4)])

Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

### 18. Given the following python code, what would the output of the code give?

my_tuppy = (1,2,5,8)

my_tuppy[2] = 6

In [130]:
my_tuppy = (1, 2, 5, 8)
my_tuppy[2] = 6

TypeError: 'tuple' object does not support item assignment

### 19. Which of the following is a python inbuilt module?

In [131]:
print('math')

math


### 20. Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?

In [136]:
df.groupby('Element')['Y2017'].sum()

Element
Domestic supply quantity                  2088198.10
Export Quantity                            182338.80
Fat supply quantity (g/capita/day)          10253.84
Feed                                       223705.68
Food                                      1258888.28
Food supply (kcal/capita/day)              454681.00
Food supply quantity (kg/capita/yr)         48690.04
Import Quantity                            294559.09
Losses                                     160614.00
Other uses (non-food)                       91645.97
Processing                                 292836.00
Production                                2030056.89
Protein supply quantity (g/capita/day)      11842.45
Residuals                                   35500.00
Seed                                        24870.14
Stock Variation                             54316.91
Total Population - Both sexes             1112641.00
Tourist consumption                            91.00
Name: Y2017, dtype: float64